### ISSUE TITLE
Runtime / NullPointerException: Cannot load from object array because "<local1>" is null

### ISSUE THREAD

**Issue by NorbertSandor** [Kotlin/kotlinx.serialization#2286]
My project is fairly complex: a class is generated by a KSP processor (based on an annotated interface), containing nested `@Serializable` 'data' classes, having different number of primary constructor properties.
There is an optimization for these generated nested classes, namely:
1. if the class has no primary constructor parameter then an `object` is generated
2. if the class has 1 primary constructor parameter then a `value class` is generated
3. if the class has multiple primary constructor parameters then a `data class` is generated

Serialization seems to work without problem but deserialization may result in a runtime `NullPointerException`, see below. (Unfortunately the (de)serialization process involves wrapping the above mentioned generated classes in a generic wrapper class which holds some common metadata besides the actual data, making the whole process more complex.)

However if I remove optimization#2, namely if in case of 1 primary constructor property a `data class` is generated instead of a `value class`, the problem goes away.
So it is somehow related to deserializing nested value classes wrapped by a generic wrapper class...

Environment:
 - Kotlin version: 1.8.20
 - Library version: 1.5.0
 - Kotlin platforms: common + js + jvm + native
 - Gradle version: 8.1.1

I could not create a reproducer until now but maybe the stacktrace provides some insights for the core developers:

```
Exception in thread "main" java.lang.NullPointerException: Cannot load from object array because "<local1>" is null
	at com.example.SomeClass$Nested_Class_cid0$$serializer.childSerializers(SomeClass.kt:52)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor$childSerializers$2.invoke(PluginGeneratedSerialDescriptor.kt:36)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor$childSerializers$2.invoke(PluginGeneratedSerialDescriptor.kt:36)
	at kotlin.SafePublicationLazyImpl.getValue(LazyJVM.kt:107)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor.getChildSerializers(PluginGeneratedSerialDescriptor.kt:36)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor.getElementDescriptor(PluginGeneratedSerialDescriptor.kt:76)
	at kotlinx.serialization.descriptors.SerialDescriptorKt$elementDescriptors$1$1.next(SerialDescriptor.kt:298)
	at kotlinx.serialization.descriptors.SerialDescriptorKt$elementDescriptors$1$1.next(SerialDescriptor.kt:293)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptorKt.hashCodeImpl(PluginGeneratedSerialDescriptor.kt:137)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor$_hashCode$2.invoke(PluginGeneratedSerialDescriptor.kt:44)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor$_hashCode$2.invoke(PluginGeneratedSerialDescriptor.kt:44)
	at kotlin.SafePublicationLazyImpl.getValue(LazyJVM.kt:107)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor.get_hashCode(PluginGeneratedSerialDescriptor.kt:44)
	at kotlinx.serialization.internal.PluginGeneratedSerialDescriptor.hashCode(PluginGeneratedSerialDescriptor.kt:97)
	at kotlinx.serialization.internal.InlineClassDescriptor.hashCode(InlineClassDescriptor.kt:20)
	at java.base/java.util.HashMap.hash(HashMap.java:338)
	at java.base/java.util.HashMap.getNode(HashMap.java:568)
	at java.base/java.util.HashMap.containsKey(HashMap.java:594)
	at java.base/java.util.HashSet.contains(HashSet.java:205)
	at kotlinx.serialization.json.internal.StreamingJsonEncoderKt.isUnsignedNumber(StreamingJsonEncoder.kt:24)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeInline(StreamingJsonDecoder.kt:351)
	at com.example.SomeClass$Nested_Class_cid0$$serializer.deserialize-a-uP6Zs(SomeClass.kt:52)
	at com.example.SomeClass$Nested_Class_cid0$$serializer.deserialize(SomeClass.kt:52)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70)
	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:162)
	at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:533)
	at com.example.SomeSerializer.deserialize(SomeSerializer.kt:48)
	at com.example.SomeSerializer.deserialize(SomeSerializer.kt:27)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70)
	at kotlinx.serialization.json.Json.decodeFromString(Json.kt:97)
```

**Comment by Prototik** [comment]
That's a bug [KT-57647](https://youtrack.jetbrains.com/issue/KT-57647/Serialization-IllegalAccessError-Update-to-static-final-field-caused-by-serializable-value-class) that was fixed in Kotlin 1.8.21

**Comment by NorbertSandor** [comment]
Really, upgrading to Kotlin 1.8.21 fixed the issue. Thanks.
